test/core/client_channel/lb_policy/round_robin_test.cc

```diff
@@ -0,0 +1,251 @@
+        << location.file() << ":" << location.line();
+    it->second.SetConnectivityState(new_state, status);
+  }
+
+  bool ConnectionRequested(size_t subchannel) {
+    auto it = subchannel_pool_.find(
+        {MakeAddress(UriForSubchannel(subchannel)), ChannelArgs()});
+    return it->second.ConnectionRequested();
+  }
+
+  std::string UriForSubchannel(size_t index) {
+    std::stringstream out;
+    out << "ipv4:127.0.0.1:44" << (index + 1);
+    return out.str();
+  }
+};
+
+TEST_F(RoundRobinTest, SingleChannel) {
+  auto policy = RoundRobinPolicy(1);
+  // LB policy should have reported CONNECTING state.
+  auto picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+  // LB policy should have requested a connection on this subchannel.
+  EXPECT_TRUE(ConnectionRequested(0));
+
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_CONNECTING);
+  picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+
+  // Channel is ready
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  // Picker should return the same subchannel repeatedly.
+  for (size_t i = 0; i < 3; ++i) {
+    ExpectPickComplete(picker.get(), UriForSubchannel(0));
+  }
+  ExpectNoStateChange();
+
+  // There's a failure
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectReresolutionRequest();
+  picker = ExpectState(GRPC_CHANNEL_TRANSIENT_FAILURE);
+  ExpectPickFail(picker.get());
+
+  // ... and a recovery!
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  ExpectPickComplete(picker.get(), UriForSubchannel(0));
+
+  ExpectNoStateChange();
+}
+
+TEST_F(RoundRobinTest, ThreeSubchannels) {
+  auto policy = RoundRobinPolicy(3);
+  // LB policy should have reported CONNECTING state.
+  auto picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  ExpectPickQueued(picker.get());
+  // LB policy should have requested a connection on this subchannel.
+  EXPECT_TRUE(ConnectionRequested(0));
+
+  for (size_t i = 0; i < 3; ++i) {
+    SetSubchannelConnectivityState(i, GRPC_CHANNEL_CONNECTING);
+    picker = ExpectState(GRPC_CHANNEL_CONNECTING);
+  }
+  ExpectPickQueued(picker.get());
+
+  // Only one channel is ready
+  SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+  // Picker should return the same subchannel repeatedly.
+  for (size_t i = 0; i < 3; ++i) {
+    ExpectPickComplete(picker.get(), UriForSubchannel(0));
+  }
+  ExpectNoStateChange();
+
+  // All channels ready
+  SetSubchannelConnectivityState(1, GRPC_CHANNEL_READY);
+  SetSubchannelConnectivityState(2, GRPC_CHANNEL_READY);
+  ExpectState(GRPC_CHANNEL_READY);
+  picker = ExpectState(GRPC_CHANNEL_READY);
+
+  std::unordered_set<std::string> reportedUris;
+  // Picker should return each address once, we do not care about the order.
+  for (size_t i = 0; i < 20; ++i) {
```

Comment:
Here we are doing 20 picks, but we are adding the address of each pick to `reportedUris` just once, so we're not actually checking that we're getting each address back the same number of times or that they're in the right order.  Instead, I suggest something like this:

```
void ExpectRoundRobinPicks(
    LoadBalancingPolicy::SubchannelPicker* picker,
    const std::vector<std::string>& addresses,
    SourceLocation location = SourceLocation()) {
  auto find_address_index = [&](const std::string& address) -> absl::optional<size_t> {
    for (size_t i = 0; i < addresses.size(); ++i) {
      if (address == addresses[i]) return i;
    }
    return absl::nullopt;
  };
  // Make 20 passes across all addresses.
  absl::optional<size_t> last_index;
  for (size_t i = 0; i < 20 * addresses.size(); ++i) {
    auto address = GetPickedAddress(picker.get());
    ASSERT_TRUE(address.has_value())
        << location.file() << ":" << location.line();
    auto address_index = find_address_index(*address);
    ASSERT_TRUE(address_index.has_value())
        << location.file() << ":" << location.line();
    if (!last_address.has_value()) {
      last_address = address_index;
    } else {
      EXPECT_EQ(*address_index, (*last_index + 1) % addresses.size())
          << location.file() << ":" << location.line();
      last_index = address_index;
    }
  }
}
```

With that helper function, individual tests can do something like this:

```
// First subchannel reports READY.  Picker always chooses that subchannel.
SetSubchannelConnectivityState(0, GRPC_CHANNEL_READY);
picker = ExpectState(GRPC_CHANNEL_READY);
ExpectRoundRobinPicks(picker.get(), {address0});
// Second subchannel reports READY.  Picker should round-robin across the two READY subchannels.
SetSubchannelConnectivityState(1, GRPC_CHANNEL_READY);
picker = ExpectState(GRPC_CHANNEL_READY);
ExpectRoundRobinPicks(picker.get(), {address0, address1});
// Last subchannel reports READY.  Picker should round-robin across all three subchannels.
SetSubchannelConnectivityState(2, GRPC_CHANNEL_READY);
picker = ExpectState(GRPC_CHANNEL_READY);
ExpectRoundRobinPicks(picker.get(), {address0, address1, address2});
```

Reply:
I added a check to make sure each channel is called the same number of times. I specifically did not check the pick order as my understanding it is not a part of a public contract (i.e. I see the randomization of the first pick was added later).

Reply:
Done.